File: app_local/lib/utils/voice_parser.dart
class VoiceParseResult {
  final String label;
  final String? color;
  final List<String> tags;

  const VoiceParseResult({
    required this.label,
    this.color,
    this.tags = const [],
  });
}

VoiceParseResult parseVoiceText(String text) {
  final raw = text.trim();
  if (raw.isEmpty) {
    return const VoiceParseResult(label: '');
  }

  const colorMap = {
    '红色': '红',
    '绿色': '绿',
    '蓝色': '蓝',
    '黑色': '黑',
    '白色': '白',
    '黄色': '黄',
    '粉色': '粉',
    '紫色': '紫',
    '灰色': '灰',
    '棕色': '棕',
    '橙色': '橙',
  };

  String? color;
  final tags = <String>{};

  for (final entry in colorMap.entries) {
    if (raw.contains(entry.key) || raw.contains(entry.value)) {
      color = entry.key;
      tags.add(entry.key);
      tags.add(entry.value);
      break;
    }
  }

  final deParts = raw.split('的').where((p) => p.trim().isNotEmpty).toList();
  var label = deParts.isNotEmpty ? deParts.last.trim() : raw;
  label = label.replaceAll(RegExp(r'^(有一?件?|有一?个?)'), '').trim();

  final colorItem = RegExp(r'(.+色)的(.+)').firstMatch(label);
  if (colorItem != null) {
    color ??= colorItem.group(1);
    label = colorItem.group(2)!.trim();
  } else if (color != null && label.startsWith(color)) {
    label = label.substring(color.length).replaceFirst(RegExp(r'^的?'), '').trim();
  }

  if (label.isEmpty) label = raw;

  tags.add(label);
  for (final part in deParts) {
    final p = part.trim();
    if (p.isNotEmpty && p.length <= 12) tags.add(p);
  }

  final keywords = RegExp(r'[\u4e00-\u9fff]{2,6}')
      .allMatches(raw)
      .map((m) => m.group(0)!)
      .where((w) => !{'有一件', '有一个', '一件', '一个'}.contains(w));
  tags.addAll(keywords);

  return VoiceParseResult(
    label: label.length > 50 ? label.substring(0, 50) : label,
    color: color,
    tags: tags.where((t) => t.length >= 2).take(8).toList(),
  );
}

File: backend/app/utils/voice_parser.py
import re
from dataclasses import dataclass


@dataclass
class VoiceParseResult:
    label: str
    color: str | None = None
    tags: list[str] | None = None


COLOR_MAP = {
    "红色": "红",
    "绿色": "绿",
    "蓝色": "蓝",
    "黑色": "黑",
    "白色": "白",
    "黄色": "黄",
    "粉色": "粉",
    "紫色": "紫",
    "灰色": "灰",
    "棕色": "棕",
    "橙色": "橙",
}


def parse_voice_text(text: str) -> VoiceParseResult:
    raw = (text or "").strip()
    if not raw:
        return VoiceParseResult(label="")

    color: str | None = None
    tags: set[str] = set()

    for full, short in COLOR_MAP.items():
        if full in raw or short in raw:
            color = full
            tags.update({full, short})
            break

    de_parts = [p.strip() for p in raw.split("的") if p.strip()]
    label = de_parts[-1] if de_parts else raw
    label = re.sub(r"^(有一?件?|有一?个?)", "", label).strip()

    color_item = re.match(r"(.+色)的(.+)", label)
    if color_item:
        color = color or color_item.group(1)
        label = color_item.group(2).strip()
    elif color and label.startswith(color):
        label = label[len(color) :].lstrip("的").strip()

    if not label:
        label = raw

    tags.add(label)
    for part in de_parts:
        if part and len(part) <= 12:
            tags.add(part)

    for m in re.finditer(r"[\u4e00-\u9fff]{2,6}", raw):
        word = m.group(0)
        if word not in {"有一件", "有一个", "一件", "一个"}:
            tags.add(word)

    clean_tags = [t for t in tags if len(t) >= 2][:8]
    return VoiceParseResult(
        label=label[:50],
        color=color,
        tags=clean_tags,
    )
